lightning/src/routing/router.rs

```diff
@@ -1128,43 +1157,27 @@ where L::Target: Logger {
 				let mut aggregate_next_hops_cltv_delta: u32 = 0;
 
 				for (idx, (hop, prev_hop_id)) in hop_iter.zip(prev_hop_iter).enumerate() {
-					// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
-					// really sucks, cause we're gonna need that eventually.
-					let hop_htlc_minimum_msat: u64 = hop.htlc_minimum_msat.unwrap_or(0);
-
-					let directional_info = DummyDirectionalChannelInfo {
-						cltv_expiry_delta: hop.cltv_expiry_delta as u32,
-						htlc_minimum_msat: hop_htlc_minimum_msat,
-						htlc_maximum_msat: hop.htlc_maximum_msat,
-						fees: hop.fees,
-					};
-
-					// We want a value of final_value_msat * ROUTE_CAPACITY_PROVISION_FACTOR but we
-					// need it to increment at each hop by the fee charged at later hops. Further,
-					// we need to ensure we round up when we divide to get satoshis.
-					let channel_cap_msat = final_value_msat
-						.checked_mul(ROUTE_CAPACITY_PROVISION_FACTOR).and_then(|v| v.checked_add(aggregate_next_hops_fee_msat))
-						.unwrap_or(u64::max_value());
-					let channel_cap_sat = match channel_cap_msat.checked_add(999) {
-						None => break, // We overflowed above, just ignore this route hint
-						Some(val) => Some(val / 1000),
-					};
-
-					let src_node_id = NodeId::from_pubkey(&hop.src_node_id);
-					let dest_node_id = NodeId::from_pubkey(&prev_hop_id);
+					let source = NodeId::from_pubkey(&hop.src_node_id);
+					let target = NodeId::from_pubkey(&prev_hop_id);
+					let candidate = network_channels
+						.get(&hop.short_channel_id)
+						.and_then(|channel| channel.as_directed_to(&target))
+						.and_then(|channel| channel.direction().map(|_| channel))
```

Comment:
If `direction()` is None, could that result in making `candidate` a `PrivateHop`?

Reply:
Yes, though I believe that would be desired given the constraint noted in `PublicHop::info`'s docs. The code here is essentially to treat a route hint as a `PrivateHop` unless we have enough information from the `NetworkGraph` to provide a better picture. Without `direction` we don't know the `cltv_expiry_delta`. I suppose we could glean that from the route hint, which would require a fourth `CandidateRouteHop` variant as a composite of `PublicHop` and `PrivateHop`'s data.